proposals/archives/java_agents_and_more_layers.md

```diff
@@ -0,0 +1,292 @@
+
+Implemented in: **v0.9.5**
+
+## Motivation
+
+There are 3 feature requests that may be able to be solved together. These are:
+
+- Give users ability to add Java agents ([#378](/../../issues/378))
+- Give users ability to copy arbitrary files to the image ([#213](/../../issues/213))
+- Separate frequently changing from non-changing dependencies for more incrementality ([#403](/../../issues/403))
+
+## Synopsis of the issues
+
+### Give users ability to add Java agents ([#378](/../../issues/378))
+
+Currently, Jib only adds application files to the container image. These include dependency JARs, resource files, and classes. However, Java developers often run their applications with Java Agents that execute as part of the JVM invocation.
+
+For example, in [this Dockerfile example](https://github.com/saturnism/spring-petclinic-gcp/blob/master/docker/Dockerfile), the Cloud Debugger and Cloud Profiler agents are added to the container image. The agent archives are first downloaded, and then extracted to their own directory on the image. The archives contains the `.so` file to pass to the java invocation in the form of:
+
+```shell
+java -agentpath:/path/to/agent/files/someagent.so ...
+```
+
+### Give users ability to copy arbitrary files to the image ([#213](/../../issues/213))
+
+Users may wish to add other files for use in the image. Currently, the way to do this would be to place the files within the application resources. However, this conflates the classpath of the application, the file can only be reached under the `/app/resources` path, and changes to the file would mean repushing all the resources. Therefore, users should have some way of adding files to a new custom layer.
+
+### Separate frequently changing from non-changing dependencies for more incrementality ([#403](/../../issues/403))
+
+The dependencies layer tends to be the largest layer in the image, since it contains all the dependency JARs. However, not all dependencies are the same. Some may change more frequently than others (especially `-SNAPSHOT` versions). Therefore, users should have some way to group different dependencies into different layers. For example, separating released artifacts from snapshot artifacts, and grouping dependencies from a BOM into a layer, or separating more frequently changed ones from less frequently changed ones to reduce the amortized push time.
+
+## Proposal
+
+The proposal is to define a new convention for adding more files to the Jib container.
+
+The user can add arbitrary files to the image by placing them in a `src/main/jib` directory. This will copy all files within `src/main/jib` to the image's root directory `/`, maintaining the same structure. For example, if the user has a text file at `src/main/jib/dir/hello.txt`, then the built image have `/dir/hello.txt`.
+
+The directory should also be able to be configured to override the `src/main/jib` default.
+
+### Alternative (Rejected) Proposal
```

Comment:
Would be good to have a `## Alternative Proposals` here as they otherwise fall under the `## Proposal`.